### PR TITLE
HPCC-13388 Documentation for how to understand workunits

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday gavin.halliday@lexisnexis.com

### CODE REVIEW

**Comment by jamienoss** [comment]
*c.f.* comments at https://github.com/ghalliday/HPCC-Platform/pull/11#issuecomment-97377636
and those yet to be considered at https://github.com/ghalliday/HPCC-Platform/commit/694aedbb7a85feb0e9f66c7515b674ea85e99290

**Comment by ghalliday** [comment]
This document is aimed at someone who is new to the system, who wants to get a broad understanding of workunits, and a reasonably detailed overview of the process of executing a query.

This is probably easier to read and review by looking directly at the rendered file:
https://github.com/ghalliday/HPCC-Platform/blob/issue13388/ecl/eclcc/WORKUNITS.rst

Any comments welcome.

@RussWhitehead you said you were interested. Please can you review it, particularly looking for gaps in the information, bits that do not make sense etc.
@afishbeck I have mentioned manifests at approximately line 310.  Please can you provide some information on what they are and when they're used.
@jakesmith please review.

**Comment by ghalliday** [comment]
A few general questions
- How should we capitalise the engines?  hThor, Thor, Roxie, or all lower case - hthor, thor, roxie.  Should we make a distinction between the engine name and the executable name?
- Are there terms that should be highlighted in bold/italics?

And most important...
- Should I adopt UK English usage and not have a comma after e.g.

**Comment by ghalliday** [comment]
btw I have removed the commas for the moment, but there is a similar question over spellings that vary, e.g. optimise v optimize etc..  Should we care?

**Comment by richardkchapman** [comment]
Generally the documentation is written in US english, but for internal documentation, comments, etc I don't really mind - so long as the meaning is clear.

**Comment by richardkchapman** [comment]
@shamser Please review

**Comment by ghalliday** [comment]
@shamser did you have any comments?  

(Normally we annotate the pull request as we find issues/questions, even if they're subsequently clarified later in the commit.)

**Comment by ghalliday** [comment]
@shamser if you have finished reviewing the document, then please include a reference to my id, and that transfers notional ownership to me.  I will then be notified that I need to look at it again.

**Comment by shamser** [comment]
Overall, very useful documentation.  I've completed my review.. Thanks @ghalliday

**Comment by ghalliday** [comment]
@shamser for future reference, github generally works better if you comment on the "files changed" rather than the individual commits.

**Comment by ghalliday** [comment]
I agree dll isn't a great term - especially since most of our audience is linux based.  I'm not sure what would be better
"Shared Library" could include a static library.  "Shared Object" might be better - since that is the general linux term.  Dynamic Library is descriptive of what it is.  @richardkchapman any votes?

**Comment by richardkchapman** [comment]
They are called dylib's on OSX. I think DLL is ok so long as we spell out what it stands for for somewhere...

**Comment by richardkchapman** [comment]
@shamser Please rereview

File: ecl/eclcc/WORKUNITS.rst [hunk 1]
```diff
@@ -0,0 +1,931 @@
...
+
+Other statistics and timings created when running the query are stored in the runtime copy of the workunit.
+(Statistics for graph elements are stored in a different format from global statistics, but the IWorkUnit interface
+ensures the implementation details are hidden.)
+
+Manifests
+---------
+
+It is possible to include other user-defined resources in the workunit dll - e.g. web pages, or dashboard layouts.
+I have to confess I do not understand them... ??Tony please provide some more information....!
```
Comment (shamser) [inline ecl/eclcc/WORKUNITS.rst]: Further information regarding user-defined resources should be provided or Line 317 should be removed.

File: ecl/eclcc/WORKUNITS.rst [hunk 2]
```diff
@@ -0,0 +1,931 @@
...
+    Provides access to functions that cannot be called inside a graph - i.e. can only be called from the global workflow code.  Most functions
+    are related to the internal implementation of particular workflow item types (e.g. persists).
+
+Glossary
+--------
+
+superfile
+    A composite file which allows a collection of files to be treated as a single compound file.
+
+?What else should go here?
```
Comment (shamser) [inline ecl/eclcc/WORKUNITS.rst]: The glossary section is a bit light but I can't thing of anything else that really 'should' be added here. The following is already covered in the document but may be briefly described here too: graph, workunit, Dali, Casandra,

**Comment by shamser** [comment]
@ghalliday Final two comments added.

File: ecl/eclcc/WORKUNITS.rst (hunk 2 above)
Comment (richardkchapman) [inline ecl/eclcc/WORKUNITS.rst]: Perhaps DLL ?

**Pull request merged**